core/consensus/babe/impl/babe_config_repository_impl.cpp

```diff
@@ -36,38 +53,744 @@ namespace kagome::consensus::babe {
+          continue;
+        }
+
+        auto saved_state_res = scale::decode<std::shared_ptr<BabeConfigNode>>(
+            encoded_saved_state_opt.value());
+
+        if (saved_state_res.has_error()) {
+          SL_WARN(logger_,
+                  "Can not decode state saved on block {}: {}",
+                  block_number,
+                  saved_state_res.error());
+          std::ignore = persistent_storage_->remove(
+              storage::kBabeConfigRepoStateLookupKey(block_number));
+          continue;
+        }
+
+        root_ = std::move(saved_state_res.value());
+        SL_DEBUG(logger_,
+                 "State was initialized by savepoint on block {}",
+                 root_->block);
+        break;
+      }
+    }
+
+    // 3. Load state from genesis, if state is still not found
+    if (root_ == nullptr) {
+      auto babe_config_res = babe_api_->configuration(genesis_block_hash_);
+      if (babe_config_res.has_error()) {
+        SL_WARN(logger_,
+                "Can't get babe config over babe API on genesis block: {}",
+                babe_config_res.error());
+        return babe_config_res.as_failure();
+      }
+      const auto &babe_config = babe_config_res.value();
+
+      root_ = BabeConfigNode::createAsRoot(
+          {0, genesis_block_hash_},
+          std::make_shared<const primitives::BabeConfiguration>(babe_config));
+      SL_DEBUG(logger_, "State was initialized by genesis block");
+    }
+
+    BOOST_ASSERT_MSG(root_ != nullptr, "The root must be initialized by now");
+
+    // Init slot duration and epoch length
+    auto slot_duration = std::chrono::duration_cast<std::chrono::milliseconds>(
+        root_->config->slot_duration);
+    BOOST_ASSERT_MSG(slot_duration.count() > 0,
+                     "Slot duration must be greater zero");
+    const_cast<BabeDuration &>(slot_duration_) = slot_duration;
+    auto epoch_length = root_->config->epoch_length;
+    BOOST_ASSERT_MSG(epoch_length, "Epoch length must be greater zero");
+    const_cast<EpochLength &>(epoch_length_) = epoch_length;
+
+    // 4. Apply digests before last finalized
+    bool need_to_save = false;
+    for (auto block_number = root_->block.number + 1;
+         block_number <= finalized_block.number;
+         ++block_number) {
+      auto block_header_res = block_tree_->getBlockHeader(block_number);
+      if (block_header_res.has_error()) {
+        SL_WARN(logger_,
+                "Can't get header of some finalized block: {}",
+                block_header_res.error());
+        return block_header_res.as_failure();
+      }
+      const auto &block_header = block_header_res.value();
+      // TODO(xDimon): Would be more efficient to take parent hash of next block
+      auto block_hash =
+          hasher_->blake2b_256(scale::encode(block_header).value());
+      primitives::BlockInfo block_info(block_number, block_hash);
+
+      for (auto &item : block_header.digest) {
+        auto res = visit_in_place(
```

Comment:
Does this duplicate digest tracker?

Reply:
No. It just variant nested in other variant